### PR TITLE
Support ring handler on lifecycler

### PR DESCRIPTION
**Pull request by simonswine**
**What this PR does**:

This PR exposes the ring introspection handlers to the lifecycler. This allow e.g. the ingester ring to be exposed on the ingester without requiring to watch the ingester ring.

**Which issue(s) this PR fixes**:

**Checklist**
- [ ] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: ring/ring.go [hunk 1]
```diff
@@ -513,27 +515,26 @@ func (r *Ring) GetReplicationSetForOperation(op Operation) (ReplicationSet, erro
 }
 
 // countTokens returns the number of tokens and tokens within the range for each instance.
-// The ring read lock must be already taken when calling this function.
-func (r *Ring) countTokens() (map[string]uint32, map[string]uint32) {
+func countTokens(ringDesc *Desc, ringTokens []uint32, ringInstanceByToken map[uint32]instanceInfo) (map[string]uint32, map[string]uint32) {
```
Comment (pstibrany) [inline ring/ring.go:518]: Shall we make this a method on `*Desc`, without extra arguments? Calling `desc.GetTokens()` and `ringDesc.getTokensInfo()` won't be as efficient, but I don't think we need to aim for efficiency when displaying web page.

**Review by pstibrany — COMMENTED** [review]
I left some minor comments, but overall this approach looks good to me. I think it's worth adding changelog entry too about Lifecycler being an HTTP handler now.

We may want to make this work on BasicLifecycler too, as there is ongoing work to switch ingesters to use that.

File: ring/http.go [hunk 2]
```diff
@@ -121,11 +122,30 @@ type httpResponse struct {
 	ShowTokens bool           `json:"-"`
 }
 
-func (r *Ring) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+type ringObserver interface {
```
[the remark this answers is not in the document]
Comment (simonswine) [reply ring/http.go]: Agreed, done within 816b94b

File: ring/ring.go (hunk 1 above)
Comment (simonswine) [reply ring/ring.go]: Agreed, done within 816b94b

File: ring/http.go [hunk 3]
```diff
@@ -121,11 +122,30 @@ type httpResponse struct {
...
+		heartbeatPeriod: heartbeatPeriod,
+	}
+}
+
+func (h *ringPageHandler) handle(w http.ResponseWriter, req *http.Request) {
 	if req.Method == http.MethodPost {
 		ingesterID := req.FormValue("forget")
-		if err := r.forget(req.Context(), ingesterID); err != nil {
-			level.Error(r.logger).Log("msg", "error forgetting instance", "err", err)
+		if err := h.forget(req.Context(), ingesterID); err != nil {
```
[the remark this answers is not in the document]
Comment (simonswine) [reply ring/http.go:129]: Good point. I have removed it in 13169e40970e57ffd7c2cf7f0c5162977a3cc5de.

I guess there is no riks of exposing additional information to the HTTP caller, as the caller has to have quite a lot of access to the cluster anyhow, if they are able to remove ingesters.

**Comment by simonswine** [comment]
> We may want to make this work on BasicLifecycler too, as there is ongoing work to switch ingesters to use that.

Done that

**Pull request merged**